File: app/app-swap/src/swap.c
#include <pb_common.h>
#include <pb_decode.h>

#include "currency_config.h"
#include "swap.h"
#include "swap.pb.h"

static bool verify_tx_signature(const uint8_t *raw_pubkey,
                                const uint8_t *tx,
                                const size_t tx_size,
                                const uint8_t *signature,
                                size_t signature_size)
{
    uint8_t hash[32];
    sha256(tx, tx_size, hash);

    cx_ecfp_public_key_t pubkey;
    ecfp_init_public_key(CX_CURVE_256K1, raw_pubkey, &pubkey);

    return ecdsa_verify(&pubkey, hash, signature, signature_size);
}

const char *handle_init_swap(const RequestInitSwap *req,
                             ResponseInitSwap *response,
                             swap_ctx_t *ctx)
{
    for (size_t i = 0; i < sizeof(ctx->device_id_swap) - 1; ++i) {
        uint8_t c;
        get_random_bytes(&c, sizeof(c));
#ifdef TESTING
        c = 42;
#endif
        ctx->device_id_swap[i] = (uint8_t)((int)'A' + c % 26);
    }

    ctx->device_id_swap[sizeof(ctx->device_id_swap) - 1] = '\x00';

    memcpy(response->device_id, ctx->device_id_swap, sizeof(response->device_id));

    return NULL;
}

static bool is_id_valid(swap_ctx_t *ctx)
{
    for (size_t i = 0; i < sizeof(ctx->device_id_swap) - 1; i++) {
        if (ctx->device_id_swap[i] == '\x00') {
            return false;
        }
    }

    return true;
}

const char *handle_swap(const RequestSwap *req, ResponseSwap *response, swap_ctx_t *ctx)
{
    if (!verify_partner(&req->partner)) {
        return "invalid partner";
    }

    ledger_swap_NewTransactionResponse tx;
    pb_istream_t stream = pb_istream_from_buffer(req->pb_tx.bytes, req->pb_tx.size);
    if (!pb_decode(&stream, ledger_swap_NewTransactionResponse_fields, &tx)) {
        return "failed to decode swap tx";
    }

    if (!is_id_valid(ctx)) {
        return "device tx id unset";
    }

    if (memcmp(ctx->device_id_swap, tx.device_transaction_id, sizeof(ctx->device_id_swap)) != 0) {
        return "device tx id doesn't match";
    }

    if (!verify_tx_signature(req->partner.pubkey, req->pb_tx.bytes, req->pb_tx.size,
                             req->signature.bytes, req->signature.size)) {
        return "invalid tx signature";
    }

    const swap_config_t *from = get_config(tx.currency_from);
    const swap_config_t *to = get_config(tx.currency_to);

    if (from == NULL) {
        return "invalid currency from";
    }

    if (to == NULL) {
        return "invalid currency to";
    }

    /* validate payout (ETH) address */
    if (!to->validate_address(tx.payout_address, req->payout_path, req->payout_path_count)) {
        return "invalid payout address";
    }

    /* validate refund (BTC) address */
    if (!from->validate_address(tx.refund_address, req->refund_path, req->refund_path_count)) {
        return "invalid refund address";
    }

    /* get printable send amount (BTC) */
    char send_amount[32];
    if (!from->get_printable_amount(tx.amount_to_provider.bytes, tx.amount_to_provider.size,
                                    send_amount, sizeof(send_amount))) {
        return "invalid send amount";
    }

    /* get printable receive amount (ETH) */
    char recv_amount[32];
    if (!to->get_printable_amount(tx.amount_to_wallet.bytes, tx.amount_to_wallet.size, recv_amount,
                                  sizeof(recv_amount))) {
        return "invalid recv amount";
    }

    /* get printable fees (BTC) */
    char fees[32];
    if (!from->get_printable_amount(req->fee.bytes, req->fee.size, fees, sizeof(fees))) {
        return "failed to get printable fees";
    }

    if (!ui_sign_tx_validation(send_amount, recv_amount, fees)) {
        return "not approved";
    }

    /* XXX */
    if (from->create_tx == NULL) {
        return NULL;
    }

    /* generate tx (BTC) */
    size_t tx_size = from->create_tx(req->refund_path, req->refund_path_count, tx.payin_address,
                                     tx.amount_to_provider.bytes, tx.amount_to_provider.size,
                                     req->fee.bytes, req->fee.size, response->tx.bytes,
                                     sizeof(response->tx.bytes));
    if (tx_size == 0) {
        return "failed to create tx";
    }

    response->tx.size = tx_size;

    return NULL;
}

File: app/app-swap/swap.py
#!/usr/bin/env python3

import os
import random
import sys

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives import hashes, serialization

# ensure that protobuf/message_pb2.py will be found
sys.path.append(os.path.dirname(os.path.realpath(__file__)))

sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", "..", "host"))

# apt install python3-protobuf
from protobuf import message_pb2, swap_pb2  # noqa: E402


import stream  # noqa: E402


class Partner:
    def __init__(self) -> None:
        self.name = b"partner"
        private_value = int.from_bytes([random.randint(0, 255) for _ in range(0, 32)], "big")
        self.private_key = ec.derive_private_key(private_value, ec.SECP256K1(), default_backend())

    def get_pubkey(self) -> bytes:
        format = serialization.PublicFormat.UncompressedPoint
        encoding = serialization.Encoding.X962
        return self.private_key.public_key().public_bytes(encoding=encoding, format=format)

    def ledger_sign(self) -> bytes:
        data = len(self.name).to_bytes(1, "big") + self.name + self.get_pubkey()
        private_value = int.from_bytes(b"\xb1\xed\x47\xef\x58\xf7\x82\xe2\xbc\x4d\x5a\xbe\x70\xef\x66\xd9"
                                       b"\x00\x9c\x29\x57\x96\x70\x17\x05\x44\x70\xe0\xf3\xe1\x0f\x58\x33", "big")
        private_key = ec.derive_private_key(private_value, ec.SECP256K1(), default_backend())
        return private_key.sign(data, ec.ECDSA(hashes.SHA256()))

    def sign_tx(self, tx) -> bytes:
        return self.private_key.sign(tx, ec.ECDSA(hashes.SHA256()))


class Tx:
    def __init__(self, payin, payout, refund, amount_to_wallet, amount_to_provider) -> None:
        tx = swap_pb2.NewTransactionResponse()

        tx.payin_address = payin
        tx.payout_address = payout
        tx.refund_address = refund
        tx.currency_from = "ETH"
        tx.currency_to = "BTC"
        tx.amount_to_wallet = amount_to_wallet
        tx.amount_to_provider = amount_to_provider
        tx.device_transaction_id = "QQQQQQQQQQ"

        self.tx = tx

    def __bytes__(self) -> bytes:
        return self.tx.SerializeToString()


class Swap:
    def init_swap_prepare_request(self) -> bytes:
        init_swap = message_pb2.RequestInitSwap()
        message = message_pb2.Request()
        message.init_swap.CopyFrom(init_swap)
        return message.SerializeToString()

    def init_swap_parse_response(self, data):
        response = message_pb2.Response()
        response.ParseFromString(data)
        assert response.WhichOneof("response") == "init_swap"
        print(f"device id: {response.init_swap.device_id}")
        self.device_id = response.init_swap.device_id
        return response

    def swap_prepare_request(self) -> bytes:
        payin = "0xd692Cb1346262F584D17B4B470954501f6715a82"
        payout = "bc1qwpgezdcy7g6khsald7cww42lva5g5dmasn6y2z"
        refund = "0xDad77910DbDFdE764fC21FCD4E74D71bBACA6D8D"

        partner = Partner()
        swap = message_pb2.RequestSwap()
        swap.partner.name = partner.name
        swap.partner.pubkey = partner.get_pubkey()
        swap.partner.signature = partner.ledger_sign()

        # ETH 0.00084 (840000000000000 wei)
        swap.fee = b"\x02\xfb\xf9\xbd\x9c\x80\x00"
        # BTC 1 (100000000 sat)
        amount_to_wallet = b"\x05\xf5\xe1\x00"
        # ETH 1.1234 (1123400000000000000 wei)
        amount_to_provider = b"\x0f\x97\x1e\x59\x14\xac\x80\x00"

        tx = Tx(payin, payout, refund, amount_to_wallet, amount_to_provider)
        swap.pb_tx = bytes(tx)
        swap.signature = partner.sign_tx(swap.pb_tx)

        swap.payout_path.extend([0x80000054, 0x80000000, 0x80000000, 1, 0])
        swap.refund_path.extend([0x8000002c, 0x8000003c, 0x80000000, 0, 0])

        message = message_pb2.Request()
        message.swap.CopyFrom(swap)
        return message.SerializeToString()

    def swap_parse_response(self, data):
        response = message_pb2.Response()
        response.ParseFromString(data)
        if response.WhichOneof("response") == "error":
            print(f"error: {response.error.error_msg}")
            sys.exit(1)
        assert response.WhichOneof("response") == "swap"
        print(f"response: {response.swap}")
        return response

    def version_prepare_request(self):
        version = message_pb2.RequestGetVersion()
        message = message_pb2.Request()
        message.get_version.CopyFrom(version)
        return message.SerializeToString()

    def version_parse_response(self, data):
        response = message_pb2.Response()
        response.ParseFromString(data)
        print(response)
        assert response.WhichOneof("response") == "get_version"
        print(f"version: {response.get_version.version}")
        return response


if __name__ == "__main__":
    parser = stream.get_stream_arg_parser()
    parser.add_argument("--action", default="swap", choices=["swap", "version"])
    args = parser.parse_args()

    with stream.get_streamer(args) as streamer:
        swap = Swap()

        if args.action == "swap":
            data = streamer.exchange(swap.init_swap_prepare_request())
            swap.init_swap_parse_response(data)

            data = streamer.exchange(swap.swap_prepare_request())
            swap.swap_parse_response(data)
        elif args.action == "version":
            data = streamer.exchange(swap.version_prepare_request())
            swap.version_parse_response(data)
